chain-core/src/property.rs

```diff
@@ -204,6 +222,58 @@ pub trait Serializable: Sized {
     fn serialize<W: std::io::Write>(&self, writer: W) -> Result<(), Self::Error>;
 
     fn deserialize<R: std::io::BufRead>(reader: R) -> Result<Self, Self::Error>;
+
+    /// Convenience method to serialize into a byte vector.
+    fn serialize_as_vec(&self) -> Result<Vec<u8>, Self::Error> {
+        let mut data = vec![];
+        self.serialize(&mut data)?;
+        Ok(data)
+    }
+}
+
+/// A trait representing the state of a chain at a particular point in
+/// time. The chain state is initialized from GenesisData and is
+/// updated by applying blocks.
+///
+/// The main purpose of this trait is to enable generic, efficient
+/// storage of chain states for various types of chains. To support
+/// efficient storage, ChainState requires implementations to support
+/// a "diff" method that produces a ChainStateDelta between arbitrary
+/// states. These ChainStateDeltas can be serialized and
+/// deserialized. (See also ChainStateStore in the chain-storage
+/// crate, which uses this trait.)
+pub trait ChainState: std::marker::Sized + Clone + Eq {
+    type Block: Block;
+    type GenesisData;
+    type Delta: ChainStateDelta;
+    type Error: std::error::Error;
+
+    /// Create a new chain state object from genesis data. The chain
+    /// length will be 0.
+    fn new(genesis_data: &Self::GenesisData) -> Result<Self, Self::Error>;
+
+    /// Update the chain state by applying a new 'block'. If an error
+    /// occurs, the chain state must be unchanged. If it succeeds,
+    /// then the chain length will be increased by 1 and
+    /// self.get_last_block_id() == block.id().
+    fn apply_block(&mut self, block: &Self::Block) -> Result<(), Self::Error>;
+
+    fn get_last_block_id(&self) -> <Self::Block as Block>::Id;
+
+    fn get_chain_length(&self) -> u64;
```

Comment:
usize in this case?

Reply:
No, usize is platform dependent which I don't think is desirable here.